### ISSUE TITLE
theme.json color gets overriden by core

### ISSUE THREAD

**Issue by 9-tailed** [WordPress/gutenberg#39423]
### Description

I changed the default color for H1 on my themes.json to green. That worked as expected. 
I then added an H1 to a cover block. I expected the H1 to have my custom green color. 
That was true for the editor, the H1 would show green, but in the front end, it is still black.
The cover block comes with "color: inherit;" if there is no color set in the editor directly, overriding my color choice from the themes.json.

### Step-by-step reproduction instructions

1. Add color for h1 in themes.json 
```
		"elements": {
			"h1": {
				"typography": {
					"fontFamily": "var(--wp--preset--font-family--titillium-web)",
					"fontWeight": "300",
					"lineHeight": "var(--wp--custom--typography--line-height--tiny)",
					"fontSize": "var(--wp--custom--typography--font-size--colossal)"
				},
				"color": {
					"text": "#3a8c38"
				}
			},
```
2. Add an h1 block to a cover block
This will result in a black h1 on the front and a green h1 in the editor.

3. Inspect front h1, you see it gets overridden by the cover style.css.

### Screenshots, screen recording, code snippet

![grafik](https://user-images.githubusercontent.com/66306592/158164794-f79f0969-3e75-46e8-b8f4-255682f74190.png) ![grafik](https://user-images.githubusercontent.com/66306592/158164865-fb728be6-7f1d-4df4-b9a4-6e1d9dc1c33b.png)

### Environment info

Wordpress: 5.9.2
Gutenberg: Version 12.7.2 
Theme: own block theme

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by skorasaurus** [comment]
I'm encountering the same issue; I'm also able to reproduce this in twentytwentythree; WP 6.1.1 and gutenberg trunk, built yesterday, 2022-12-01 (last commit was 4acfb3a6c5f79bd0b0c393190717eff4a340c91d) 

by: 

1. I changed the default color for H1 in global settings (not in theme.json but in the editor) That worked as expected. 
2. I then added an H1 to a cover block (yes, I know it's wrong HTML semantics but the same issue would apply if it were an h2). . I expected the H1 to have my custom green color.
The H1 within the cover block was still black both in the editor and in the front end. 

@9-tailed Could you modify the title to be more descriptive of the issue? something like: H1 color defined in global styles is overwritten in cover block?

**Comment by maffi-git** [comment]
This definately needs more work. 
color:inherit is applied to all heading-blocks h1-h6 inside a cover block that do not have a text-color set. 
I was currently wondering if this is expected behaviour or not. The cover-block has no color-settings for text or background. But it has a color-setting for its overlay. 
So we either need a color-setting for the cover-block or some other sort of control for its colors. 
The problem starts with the cover block's initial color-settings of its inner block here: 
```
.wp-block-cover-image .wp-block-cover__inner-container, .wp-block-cover .wp-block-cover__inner-container {
  color: #fff;
}
```
In my case and @skorasaurus the twentytwentythree-theme even pops in a third class: 
```
.wp-block-cover-image.is-light .wp-block-cover__inner-container, .wp-block-cover.is-light .wp-block-cover__inner-container {
  color: #000;
}
```
Basically, I see no need for the cover-block to set these color-styles, that cannot be altered in the editor. So in my opinion we should delete them all:
.wp-block-cover_inner-container should have no colors set
.wp-block-cover h1-6:not(.has-text-color) {color: inherit;} should be droped 

This way the color of headings-blocks get's set in the site-editor and no override is necessary.
Within the normal editor users can set their colors too.

**Comment by maffi-git** [comment]
See also this thread [33723](https://github.com/WordPress/gutenberg/issues/33723) where the inner-containers' color also causes problems.

**Comment by maffi-git** [comment]
And here is the explanation, why we have these inherits. [17728](https://github.com/WordPress/gutenberg/pull/17728) which brings me back to my first suggestion: cover-block should not have hard-coded colors and could do with some text-color-control. This way all blocks inside the cover block can inherit their colors, if set, from the site editor.

**Comment by TonyGravagno** [comment]
WP v6.3.2 Production : I just spent a lot of time diagnosing this and traced it down to [/wp-includes/blocks/cover/style.css#L118](https://github.com/WordPress/WordPress/blob/9982f67329e975d12c717e5510c6caca95aa727a/wp-includes/blocks/cover/style.css#L118). Then before creating a ticket I did a search and found this open issue. I'm so glad it's not just me. 😉 

To Reproduce : You can change any heading style H1-6 in a page and it will be reflected in the page. But if you set the heading style in the Global style editor, for a theme, child theme, or style variation, the change is ignored and the heading shows as \#000. In the browser developer tools, disable "color:inherit;" as shown by @9-tailed and you get your H1-H6 colors. You must further disable the next inherit to override the hard-coded \#000 and see the general "Heading" color.

### However...

<img align="right" src="https://github.com/WordPress/gutenberg/assets/499473/17355c09-d5e7-4814-ba3e-c580365dc8a2">

This specific issue is **not** due to the inherits, though that is an issue that's surfaced here.

`.wp-block-cover h1:not(.has-text-color) ... h2, h3...`
Why is that not detecting that the headers have text color? If that was working then the inherit should not be executed.

`<h3 class="wp-block-heading alignfull has-text-align-left wp-container-content-8">H3 Header</h3>`
There ya go ... no 'has-text-color' class. Add that in with dev tools, the rule fails, and you get your color.

This specific problem seems to stem from the class ID not getting injected into the tag class attributes when H1-6 styling is applied anywhere. Either that needs to be fixed, or these tests should be looking for some other marker.

And then, yes, remove all of your custom level-specific heading colors from the theme/variation and you'll still get black rather than the theme-specified heading color. 🤓

**Comment by TonyGravagno** [comment]
I've created a Gist as that fixes this issue for me, until it's addressed in core. Comments are included to explain how it works.

Please comment on the gist if you have questions or issues ... especially if your comment is something like "you idiot, this isn't the way to do this!", and you have the creds to say so. 😁 
https://gist.github.com/TonyGravagno/06f03622af89abcb7069a96ab2c93920

**Comment by webexpr-dhenriet** [comment]
@TonyGravagno, @maffi-git, @skorasaurus, @carolinan 

This seems to have been corrected for H1 headings and paragraphs, but for h2,h3,h4,h5,h6 headings it still occurs.
However, I have this in theme.json (custom hybrid theme)
```json
"elements": {
   "heading": {
        "typography": {
            "fontWeight": "600",
            "lineHeight": "1.2"
        },
        "color": {
            "text": "var(--wp-custom--secondary-color)"
        }
    },
}
```

What is applied to h1 
```css
.editor-styles-wrapper .wp-block-cover h1 {
    color: var(--wp--preset--color--white);
    font-weight: 700;
}
```

I've set the font weight to 600, only the contrast color should be corrected.
If i switch to h2, the correct font weight is applied, and the color is no longer the right one.
```css
.editor-styles-wrapper h1, .editor-styles-wrapper h2, .editor-styles-wrapper h3, .editor-styles-wrapper h4, .editor-styles-wrapper h5, .editor-styles-wrapper h6 {
    color: var(--wp-custom--secondary-color);
    font-weight: 600;
    line-height: 1.2;
}
```

**Simple fix** 

For my part, I've been doing a simple fix for a while now that I load into my admin stylesheet, the same one that applies styles with `enqueue_block_editor_assets`

```scss
.editor-styles-wrapper {

    .wp-block-cover {

        .wp-block-heading,
        .wp-block-paragraph {
            color: inherit;
        }
    }
}
```

**Optimization**

I've also noticed something that could be optimized, I'm not sure, but I suggest just in case.
When we set a color on a title or content in the admin, the style is inline `style="color:...;"` : 
```html
<h2 
  class="... has-tp-primary-color has-text-color" 
  style="white-space: pre-wrap;min-width: 1px;color: rgb(0, 255, 152);">
  Title
</h2>
```
Why do we need this rule `:where(:not(.has-text-color))` for each tag type ?

```css
.editor-styles-wrapper .has-large-font-size {
    font-size: var(--wp--preset--font-size--large) !important;
}
```

```css
//== Actually
.wp-block-cover h1:where(:not(.has-text-color)), 
.wp-block-cover h2:where(:not(.has-text-color)), 
.wp-block-cover h3:where(:not(.has-text-color)),
.wp-block-cover h4:where(:not(.has-text-color)),
.wp-block-cover h5:where(:not(.has-text-color)),
.wp-block-cover h6:where(:not(.has-text-color)),
.wp-block-cover p:where(:not(.has-text-color)),
.wp-block-cover-image h1:where(:not(. has-text-color)),
.wp-block-cover-image h2:where(:not(.has-text-color)),
.wp-block-cover-image h3:where(:not(.has-text-color)),
.wp-block-cover-image h4:where(:not(.has-text-color)), 
.wp-block-cover-image h5:where(:not(.has-text-color)), 
.wp-block-cover-image h6:where(:not(.has-text-color)), 
.wp-block-cover-image p:where(:not(.has-text-color)) {
  color: inherit;
}

//== Optimization
.wp-block-cover .wp-block-heading:where(:not(.has-text-color)),
.wp-block-cover .wp-block-paragraph:where(:not(.has-text-color)),
.wp-block-cover-image .wp-block-heading:where(:not(.has-text-color)),
.wp-block-cover-image .wp-block-paragraph:where(:not(.has-text-color))  {
    color: inherit;
}

//== Better optimization
.wp-block-cover .wp-block-heading,
.wp-block-cover .wp-block-paragraph,
.wp-block-cover-image .wp-block-heading,
.wp-block-cover-image .wp-block-paragraph  {
    color: inherit;
}
```

I'd also like to take this opportunity to ask you about something I've noticed that also overloads the titles, why is this css class applied 'has-large-font-size'?
I add a cover, by default it's a paragraph that's inserted, I click to turn it into a title, and then I get a 'has-large-font-size' class. If I add a title directly (without transforming a paragraph into a title), then I don't get this class added. 

```html
<h2 class="... has-large-font-size ...">Title</h2>
```
![Screenshot_2](https://github.com/WordPress/gutenberg/assets/62699215/6bbd2cf6-5938-4144-9f0e-60eddc11ea2c)

**Comment by skorasaurus** [comment]
@webexpr-dhenriet 

Thanks for the lengthy and detailed notes; did you have the gutenberg plugin installed when you made these observations and if so, which version?

**Comment by webexpr-dhenriet** [comment]
@skorasaurus 

I made this observations without Gutenberg plugin. 

I have two test sites, one I'm currently working on for a delivery (I called it in this case figur and a second to test anything.
I've noticed a difference between sandbox and sandbox-v2.

But I just tested with 
- twentytwentyfour without Gutenberg (sandbox)
- twentytwentyfour with Gutenberg Version 18.6.1 (sandbox)
- hybrid custom theme without Gutenberg (sandbox, sanbox-v2)
- hybrid custom theme with Gutenberg Version 18.6.1 (sandbox , sanbox-v2)

Sanbox-v2 is the theme that has the 2 problems, color on the cover title and also the font weight 'has-large-font-size' (with and without Gutenberg, not tested with twentytwentyfour).

Sandbox only has the problem with the 'has-large-font-size' font weight of the cover title, not the color (with and without Gutenberg, with twentytwentyfour, with a hybrid custom theme ). 
Installing or not installing gutenberg doesn't seem to be the problem.

In Sanbox-v2, 'color: inherit;' is loaded but overwritten by the value declared in theme.json, I've noticed that the loading of 'color: inherit;' differs between Sandbox and Sanbox-v2.
It's not loading from "wp-includes/css/dist/block-library/style.min.css".
Maybe it's the rasion, but why ...

Could it be the PHP constants? `COMPRESS_CSS`, `COMPRESS_SCRIPTS`, `CONCATENATE_SCRIPTS `
I haven't defined them in either case.

OK =>http://sandbox.local/wp-includes/css/dist/block-library/style.min.css?ver=6.5.5
KO => https://sandbox-v2.local/wp-admin/load-styles.php?c=1&dir=ltr&load%5Bchunk_0%5D=dashicons,admin-bar,buttons,media-views,editor-buttons,wp-components,wp-preferences,wp-block-editor,wp-reusable-blocks,wp-patter&load%5Bchunk_1%5D=ns,wp-editor,common,forms,wp-reset-editor-styles,wp-block-library,wp-block-editor-content,wp-edit-blocks,wp-commands,wp-edit-pos&load%5Bchunk_2%5D=t,wp-block-directory,wp-format-library,admin-menu,dashboard,list-tables,edit,revisions,media,themes,about,nav-menus,wp-pointer,w&load%5Bchunk_3%5D=idgets,site-ico

![Screenshot_2](https://github.com/WordPress/gutenberg/assets/62699215/b03c5c5b-b513-4386-b572-767fa14d64a6)

**Comment by carolinan** [comment]
> I'd also like to take this opportunity to ask you about something I've noticed that also overloads the titles, why is this css class applied 'has-large-font-size'?
I add a cover, by default it's a paragraph that's inserted, I click to turn it into a title, and then I get a 'has-large-font-size' class. If I add a title directly (without transforming a paragraph into a title), then I don't get this class added.

When you transform a paragraph into a heading it keeps its styles.
You can test this anywhere in the editor by adding a font size or color to a paragraph and changing it to the heading block.
It is a quality of life feature that users have requested. Without it, you would ned to reapply the styles after the transform.

Regarding the suggest CSS optimisation, please know that if you change `h1` etc to `wp-block-heading` it will **only** work for the heading block not any other headings like post titles, and that would be an unexpected change.
`wp-block-paragraph` does not exist. There is an open issue and PR for it.

**Comment by carolinan** [comment]
Also there is a reason why the cover block uses a paragraph and not a heading, that is because headings have semantic meaning. They should only be uses as headings describing the content below it.
When the cover block is placed, **the block** does not know what is below it, so the user will need to edit the blocks depending on the context, using the correct heading level for the page.

Please be aware of this because if it is only a matter of styling, not content formatting, then adding headings can make the page less accessible.
